### PR TITLE
fix(status): redesign env info metrics

### PR DESCRIPTION
**Pull request by oowl**
change lmdb info table entry to `used_pages` `alocated_pages` `entries`
```
    lst->used_pages     = mst.ms_branch_pages + mst.ms_leaf_pages
                          + mst.ms_overflow_pages;
    lst->alocated_pages = mei.me_last_pgno + 1;
    lst->entries        = mst.ms_entries;
```

FTI-5111

### CODE REVIEW

**Review by chronolaw — DISMISSED** [review]

File: lib/resty/lmdb/status.lua [hunk 1]
```diff
@@ -15,11 +15,11 @@ ffi.cdef([[
     typedef struct {
         size_t          map_size;        /**< Size of the data memory map */
         unsigned int    page_size;       /**< Size of a database page. */
-        size_t          max_map_size;    /**< Size of the data memory map */
-        unsigned int    last_used_page;  /**< page numbers of the last used pages */
-        size_t          last_txnid;      /**< ID of the last committed transaction */
         unsigned int    max_readers;     /**< max reader slots in the environment */
         unsigned int    num_readers;     /**< max reader slots used in the environment */
+        unsigned int    alocated_pages;  /**< number of pages allocated */
```
Comment (dndx) [inline lib/resty/lmdb/status.lua:20]:
```suggestion
        unsigned int    allocated_pages;  /**< number of pages allocated */
```

Misspelling

**Review by dndx — CHANGES_REQUESTED** [review]

**Comment by dndx** [comment]
@oowl Tests are broken

File: README.md [hunk 2]
```diff
@@ -73,13 +73,13 @@ In case of error, `nil` and a string describing the error will be returned inste
...
-  "max_readers": 126,
-  "num_readers": 1,
-  "map_size": 1048576,
-  "page_size": 4096,
-  "max_map_size": 1048576
+    "page_size": 4096,
+    "max_readers":126,
+    "num_readers": 1,
+    "allocated_pages": 2,
+    "used_pages": 0,
```
Comment (dndx) [inline README.md:80]:
```suggestion
    "in_use_pages": 0,
```

`used_pages` makes it sounds like the pages are not reusable later.

File: lib/resty/lmdb/status.lua [hunk 3]
```diff
@@ -38,11 +38,11 @@ function _M.get_env_info()
...
         map_size       = tonumber(env_status[0].map_size),
         page_size      = tonumber(env_status[0].page_size),
-        max_map_size   = tonumber(env_status[0].max_map_size),
-        last_used_page = tonumber(env_status[0].last_used_page),
-        last_txnid     = tonumber(env_status[0].last_txnid),
         max_readers    = tonumber(env_status[0].max_readers),
         num_readers    = tonumber(env_status[0].num_readers),
+        allocated_pages = tonumber(env_status[0].allocated_pages),
+        in_use_pages     = tonumber(env_status[0].in_use_pages),
+        entries        = tonumber(env_status[0].entries),
```
Comment (chronolaw) [inline lib/resty/lmdb/status.lua:45]: style issue: align `=` as above.

File: src/ngx_lua_resty_lmdb_module.c [hunk 4]
```diff
@@ -706,11 +706,12 @@ int ngx_lua_resty_lmdb_ffi_env_info(ngx_lua_resty_lmdb_ffi_status_t *lst, const
...
     lst->page_size      = mst.ms_psize;
-    lst->max_map_size   = mei.me_mapsize;
-    lst->last_used_page = mei.me_last_pgno + 1;
-    lst->last_txnid     = mei.me_last_txnid;
     lst->max_readers    = mei.me_maxreaders;
     lst->num_readers    = mei.me_numreaders;
+    lst->in_use_pages     = mst.ms_branch_pages + mst.ms_leaf_pages
+                          + mst.ms_overflow_pages;
+    lst->allocated_pages = mei.me_last_pgno + 1;
+    lst->entries        = mst.ms_entries;
```
Comment (chronolaw) [inline src/ngx_lua_resty_lmdb_module.c:714]: style issue: align `=` as above.

File: src/ngx_lua_resty_lmdb_module.h [hunk 5]
```diff
@@ -44,11 +44,11 @@ struct ngx_lua_resty_lmdb_operation_s {
...
     size_t          map_size;        /**< Size of the data memory map */
     unsigned int    page_size;       /**< Size of a database page. */
-    size_t          max_map_size;    /**< Size of the data memory map */
-    unsigned int    last_used_page;  /**< page numbers of the last used pages */
-    size_t          last_txnid;      /**< ID of the last committed transaction */
     unsigned int    max_readers;     /**< max reader slots in the environment */
     unsigned int    num_readers;     /**< max reader slots used in the environment */
+    unsigned int    allocated_pages;  /**< number of pages allocated */
+    size_t          in_use_pages;      /**< number of pages used */
+    unsigned int    entries;         /**< the number of entries (key/value pairs) in the environment */
```
Comment (chronolaw) [inline src/ngx_lua_resty_lmdb_module.h:51]: style issue: align `/*` as above.

**Review by dndx — APPROVED** [review]

**Review by dndx — APPROVED** [review]

**Review by chronolaw — APPROVED** [review]

**Comment by oowl** [comment]
It seems this implementation has some problem, `mdb_env_stat` function call return `lmdb` database page usage filed `ms_leaf_pages` is always 1, it's unexpected for our use-case. gdb debug process in here:

## Before post config to the dbless mode Kong.
```
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── code:x86:64 ────
   0x55d6b57e732c <ngx_lua_resty_lmdb_ffi_env_info+204> mov    rax, QWORD PTR [rbp-0x78]
   0x55d6b57e7330 <ngx_lua_resty_lmdb_ffi_env_info+208> mov    DWORD PTR [rax+0x8], edx
   0x55d6b57e7333 <ngx_lua_resty_lmdb_ffi_env_info+211> mov    edx, DWORD PTR [rbp-0x10]
 → 0x55d6b57e7336 <ngx_lua_resty_lmdb_ffi_env_info+214> mov    rax, QWORD PTR [rbp-0x78]
   0x55d6b57e733a <ngx_lua_resty_lmdb_ffi_env_info+218> mov    DWORD PTR [rax+0xc], edx
   0x55d6b57e733d <ngx_lua_resty_lmdb_ffi_env_info+221> mov    edx, DWORD PTR [rbp-0xc]
   0x55d6b57e7340 <ngx_lua_resty_lmdb_ffi_env_info+224> mov    rax, QWORD PTR [rbp-0x78]
   0x55d6b57e7344 <ngx_lua_resty_lmdb_ffi_env_info+228> mov    DWORD PTR [rax+0x10], edx
   0x55d6b57e7347 <ngx_lua_resty_lmdb_ffi_env_info+231> mov    rdx, QWORD PTR [rbp-0x58]
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── source:/home/ouyang/.c[...].c+31 ────
     26          return NGX_ERROR;
     27      }
     28  
     29      lst->map_size           = mei.me_mapsize;
     30      lst->page_size          = mst.ms_psize;
             // lst=0x00007ffe233cbe28  →  [...]  →  0x0000000080000000
 →   31      lst->max_readers        = mei.me_maxreaders;
     32      lst->num_readers        = mei.me_numreaders;
     33      lst->in_use_pages       = mst.ms_branch_pages + mst.ms_leaf_pages
     34                                  + mst.ms_overflow_pages;
     35      lst->allocated_pages    = mei.me_last_pgno + 1;
     36      lst->entries            = mst.ms_entries;
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── threads ────
[#0] Id 1, Name: "nginx", stopped, reason: SINGLE STEP
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── trace ────
[#0] 0x55d6b57e7336 → ngx_lua_resty_lmdb_ffi_env_info(lst=0x55d6b7c371e0, err=0x55d6b642a110)
[#1] 0x7ff3b766bfde → lj_vm_ffi_call()
[#2] 0x7ff3b76cd4e1 → lj_ccall_func(L=<optimized out>, cd=<optimized out>)
[#3] 0x7ff3b76e7925 → lj_cf_ffi_meta___call(L=0x55d6b80b9720)
[#4] 0x7ff3b7669b93 → lj_BC_FUNCC()
[#5] 0x55d6b5750c52 → ngx_http_lua_run_thread(L=0x55d6b64842d0, r=0x55d6b7babda0, ctx=0x55d6b6a73e60, nrets=0x1)
[#6] 0x55d6b5771ee8 → ngx_http_lua_sleep_resume(r=0x55d6b7babda0)
[#7] 0x55d6b5771c53 → ngx_http_lua_sleep_handler(ev=0x55d6b6a73ee0)
[#8] 0x55d6b5626c5d → ngx_event_expire_timers()
[#9] 0x55d6b5623efe → ngx_process_events_and_timers(cycle=0x55d6b594ba10)
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
gef➤  p mei
$1 = {
  me_mapaddr = 0x0,
  me_mapsize = 0x80000000,
  me_last_pgno = 0x5,  <-  focus this
  me_last_txnid = 0x2,
  me_maxreaders = 0x7e,
  me_numreaders = 0x1
}
gef➤  p msi
No symbol "msi" in current context.
gef➤  p mst
$2 = {
  ms_psize = 0x1000,
  ms_depth = 0x1,
  ms_branch_pages = 0x0,  <-  focus this
  ms_leaf_pages = 0x1, <-  focus this
  ms_overflow_pages = 0x0,  <-  focus this
  ms_entries = 0x1
}

```

## Post Config
```
╭─ouyang@home ~/work/kong/self/kong ‹bazel/debug-symbol●› 
╰─$ http 127.0.0.1:8001/config config=@~/Documents/api/kong.yml

╭─ouyang@home ~/work/kong/self/kong ‹bazel/debug-symbol●› 
╰─$ curl 127.0.0.1:8001/services | jq .data | grep id | wc -l                                                                                                                                             
15
```


## After post config to the dbless mode Kong.
```
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── code:x86:64 ────
   0x562d5d54e329 <ngx_lua_resty_lmdb_ffi_env_info+201> mov    edx, DWORD PTR [rbp-0x60]
   0x562d5d54e32c <ngx_lua_resty_lmdb_ffi_env_info+204> mov    rax, QWORD PTR [rbp-0x78]
   0x562d5d54e330 <ngx_lua_resty_lmdb_ffi_env_info+208> mov    DWORD PTR [rax+0x8], edx
 → 0x562d5d54e333 <ngx_lua_resty_lmdb_ffi_env_info+211> mov    edx, DWORD PTR [rbp-0x10]
   0x562d5d54e336 <ngx_lua_resty_lmdb_ffi_env_info+214> mov    rax, QWORD PTR [rbp-0x78]
   0x562d5d54e33a <ngx_lua_resty_lmdb_ffi_env_info+218> mov    DWORD PTR [rax+0xc], edx
   0x562d5d54e33d <ngx_lua_resty_lmdb_ffi_env_info+221> mov    edx, DWORD PTR [rbp-0xc]
   0x562d5d54e340 <ngx_lua_resty_lmdb_ffi_env_info+224> mov    rax, QWORD PTR [rbp-0x78]
   0x562d5d54e344 <ngx_lua_resty_lmdb_ffi_env_info+228> mov    DWORD PTR [rax+0x10], edx
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── source:/home/ouyang/.c[...].c+31 ────
     26          return NGX_ERROR;
     27      }
     28  
     29      lst->map_size           = mei.me_mapsize;
     30      lst->page_size          = mst.ms_psize;
             // lst=0x00007fffda19f5d8  →  [...]  →  0x0000000080000000
 →   31      lst->max_readers        = mei.me_maxreaders;
     32      lst->num_readers        = mei.me_numreaders;
     33      lst->in_use_pages       = mst.ms_branch_pages + mst.ms_leaf_pages
     34                                  + mst.ms_overflow_pages;
     35      lst->allocated_pages    = mei.me_last_pgno + 1;
     36      lst->entries            = mst.ms_entries;
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── threads ────
[#0] Id 1, Name: "nginx", stopped, reason: SINGLE STEP
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── trace ────
[#0] 0x562d5d54e333 → ngx_lua_resty_lmdb_ffi_env_info(lst=0x562d6253bbe0, err=0x562d600b7110)
[#1] 0x7f18a1a6dfde → lj_vm_ffi_call()
[#2] 0x7f18a1acf4e1 → lj_ccall_func(L=<optimized out>, cd=<optimized out>)
[#3] 0x7f18a1ae9925 → lj_cf_ffi_meta___call(L=0x562d624a52a0)
[#4] 0x7f18a1a6bb93 → lj_BC_FUNCC()
[#5] 0x562d5d4b7c52 → ngx_http_lua_run_thread(L=0x562d601112d0, r=0x562d62ea75b0, ctx=0x562d60533b50, nrets=0x1)
[#6] 0x562d5d4d8ee8 → ngx_http_lua_sleep_resume(r=0x562d62ea75b0)
[#7] 0x562d5d4d8c53 → ngx_http_lua_sleep_handler(ev=0x562d60533bd0)
[#8] 0x562d5d38dc5d → ngx_event_expire_timers()
[#9] 0x562d5d38aefe → ngx_process_events_and_timers(cycle=0x562d5f5d8a10)
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
gef➤  p mei
$1 = {
  me_mapaddr = 0x0,
  me_mapsize = 0x80000000,
  me_last_pgno = 0x27,  <-  focus this
  me_last_txnid = 0x3,
  me_maxreaders = 0x7e,
  me_numreaders = 0x1
}
gef➤  p mst
$2 = {
  ms_psize = 0x1000,
  ms_depth = 0x1,
  ms_branch_pages = 0x0,  <-  focus this
  ms_leaf_pages = 0x1,  <-  focus this
  ms_overflow_pages = 0x0,  <-  focus this
  ms_entries = 0x1
}

```

**Comment by oowl** [comment]
we can see that `me_last_pgno` is increasing, but `ms_entries` and `ms_leaf_pages`, and `ms_branch_pages` has no any change after kong config change.

**Comment by chronolaw** [comment]
Could we create a new PR to fix it?

**Comment by oowl** [comment]
> Could we create a new PR to fix it?

I have tried to debug this in LMDB internal source code, But we can not resolve this shortly. To ensure the smooth release process of version 3.4, I suggest removing the relevant PRs regarding this feature from the 3.4 task.

**Comment by dndx** [comment]
@oowl If you build using the `mdb.master` branch and revert the page encryption feature. Does the bug still exist?

We are on an experimental branch of LMDB for DB encryption and it is known to have other quality issues.

**Comment by oowl** [comment]
> @oowl If you build using the `mdb.master` branch and revert the page encryption feature. Does the bug still exist?
> 
> We are on an experimental branch of LMDB for DB encryption and it is known to have other quality issues.

I have test it in `mdb.master`, and It still has these issues. I will revert kong ce related PR.

**Pull request merged**